### ISSUE TITLE
Explicit *unambigous* type for all function arguments and `class` fields

### ISSUE THREAD

**Issue by SamuelMarks** [keras-team/keras#18817]
Following the closure of #18786 I thought maybe a discussion before opening the PR was what was missing.

My project [`cdd-python`](https://github.com/offscale/cdd-python) can take whole modules as input, and apply type hints, convert between docstring formats, or emit SQLalchemy, argparse, `class`es, or functions; e.g.,
- see this example that takes `keras.optimizers` as input and produces `class`es as output:
```sh
python -m cdd exmod -m keras.optimizers --emit class -o generated\optimizers
```

Here is the output from the https://github.com/keras-team/keras/blob/1c1c1d0/keras/optimizers/adam.py (+interpolated) input:

```py
"""
Generated from keras.src.optimizers.adam.
"""
from typing import Optional, Any


class Adam(object):
    """
    Optimizer that implements the Adam algorithm.

        Adam optimization is a stochastic gradient descent method that is based on
        adaptive estimation of first-order and second-order moments.

        According to
        [Kingma et al., 2014](http://arxiv.org/abs/1412.6980),
        the method is "*computationally
        efficient, has little memory requirement, invariant to diagonal rescaling of
        gradients, and is well suited for problems that are large in terms of
        data/parameters*".

    :cvar learning_rate: A float, a `keras.optimizers.schedules.LearningRateSchedule` instance, or a callable that takes no arguments and returns the actual value to use. The learning rate.
    :cvar beta_1: A float value or a constant float tensor, or a callable that takes no arguments and returns the actual value to use. The exponential decay rate for the 1st moment estimates.
    :cvar beta_2: A float value or a constant float tensor, or a callable that takes no arguments and returns the actual value to use. The exponential decay rate for the 2nd moment estimates.
    :cvar epsilon: A small constant for numerical stability. This epsilon is "epsilon hat" in the Kingma and Ba paper (in the formula just before Section 2.1), not the epsilon in Algorithm 1 of the paper.
    :cvar amsgrad: Boolean. Whether to apply AMSGrad variant of this algorithm from the paper "On the Convergence of Adam and beyond".
    """

    learning_rate: float = 0.001
    beta_1: float = 0.9
    beta_2: float = 0.999
    epsilon: float = 1e-07
    amsgrad: bool = False
    global_clipnorm: Optional[Any] = None
    use_ema: bool = False
    name: str = "adam"
    ema_overwrite_frequency: Optional[Any] = None
    clipnorm: Optional[Any] = None
    ema_momentum: float = 0.99
    clipvalue: Optional[Any] = None
    weight_decay: Optional[Any] = None


__all__ = ["Adam"]
```

Or another example; this time with your https://github.com/keras-team/keras/blob/b9be76a/keras/callbacks/reduce_lr_on_plateau.py †:
```py
class ReduceLROnPlateau(object):
    monitor: str = "val_loss"
    factor: float = 0.1
    patience: int = 10
    verbose: int = 0
    mode: Literal["auto", "max", "min"] = "auto"
    min_delta: float = 0.0001
    cooldown: int = 0
    min_lr: int = 0
```

For https://github.com/keras-team/keras/blob/b9be76a/keras/callbacks/backup_and_restore_callback.py †:
```py
class BackupAndRestore(object):
    backup_dir: Optional[Any]
    save_freq: Union[Literal["epoch"], bool, int] = "epoch"
    delete_checkpoint: bool = True
```

†to be succinct on this GitHub issue; just showing the `class` keyword, name, and fields.

---

As you can see, though docstring parsing (+ your adhoc type syntax parsing), and inferring from the default value type can give *some* reasonable resultant types; it is imperfect.

For example, lines like:
```
A float value or a constant float tensor, or a callable that takes no arguments and returns the actual value to use.
```

Needs to become:
```py
Union[float, keras.types.FloatTensorType, Callable[[],Union[float, keras.types.FloatTensorType]]
```

(or with (PEP604)[https://peps.python.org/pep-0604/] `|` can be used instead of `Union` and [I think] `Literal`s)

---

@fchollet can you reconsider using an unambiguous parameter type specification language, like Python's official [PEP484](https://peps.python.org/pep-0484/) and its derivatives as per the PR you rejected in #18786 ?

Thank you

**Comment by ID6109** [comment]
@SamuelMarks, my apologies for being blunt, but I fail to see how using an unambiguous parameter type specification language helps our case here. The docstrings are primarily (and quite obviously) meant to be human-readable in order to assist users, potential contributors, as well as maintainers. Someone without intricate knowledge of Python's official PEP484 would find it hard to decode what a statement such as `Union[float, keras.types.FloatTensorType, Callable[[],Union[float, keras.types.FloatTensorType]]` implies.

**Comment by fchollet** [comment]
Thanks for the suggestion. Overall, we aren't interested in adding this feature.

> The docstrings are primarily (and quite obviously) meant to be human-readable

This is True. Docstrings are for humans. More generally we've stayed away form type annotations because historically they've caused more issues for us than they've solved problems. 

> The main advantage, relevant to Keras, is for [Google Vizier](https://github.com/google/vizier) (also [Ray Tune](https://docs.ray.io/en/latest/tune/index.html); and others): know what the search space is, so these can be plugged into the search-space and optimised accordingly.

Automatically generating a hparam search space from model code is an impasse (unless you are leveraging ML to do it in a smart way informed by a strong prior based on past data). There is often no distinction between discrete values and continuous values in the API (e.g. the optimizer LR can be a float or a function; the actual range here is the space of all possible scalar functions). And there's an enormous number of possible argument to tune, the range of possible values for almost any argument is infinite, the search space grows combinatorially with number of argument being tuned, and evaluating any single combination is expensive. (For instance, it would be incredibly wasteful to evaluate all possible optimizers when the thing you should focus on is purely the LR schedule for your `AdamW` optimizer).

This is why hparam optimization relies solely on handcrafted search spaces; you need a human (or a very smart algorithm) to tell you which specific arguments to pay attention to and which narrow value range to look at. You're going to want to explore only the most import 0.01% of possible combinations.

**Comment by SamuelMarks** [comment]
> Automatically generating a hparam search space from model code is an impasse (unless you are leveraging ML to do it in a smart way informed by a strong prior based on past data).

I'm doing this partially as a chapter (or two) for my second PhD; partially for some potentially commercial (but open-source) work; and partially for my self-funded charitable medical diagnostic work (at Harvard Medical School / Mass Eye and Ear).

Yes, the idea is most certainly to do this intelligently rather than to greedily search the entire space.

---

As for future work, very interested in limiting number ranges:
- https://github.com/python/mypy/issues/16117
- https://github.com/python/typing/issues/1304#issuecomment-1722091885

Often in ML codebases [incl. Keras, JAX, TensorFlow, PyTorch, scikit-learn] there are integers that need to be nonnegative, and floating point numbers that are required to be, for example, inbetwixt `0.0` and `1.0`.

What I am trying to do to Keras—and then to those other frameworks/libraries—is to _expose_ what categorical and continuous variables are available (incl. their ranges); and then use this to **benchmark**, and also to **_expose_ interfaces** for interaction in:
- Databases (ORM, so multi-database)
- GUI (cross-platform)
- REST API
- CLI (cross-platform)
- Frontends (web, iOS, Android)
- Python SDKs
- Other language SDKs
  - writing [`cdd-c`](https://github.com/SamuelMarks/cdd-c)—like my Python compiler but written in C for C; so C would be the base and every other language can access—think [libtensorflow](https://www.tensorflow.org/install/lang_c) but wrapped in dozens of programming languages

---

PS: Some test-cases that I got working based on your adhoc syntax - https://github.com/offscale/cdd-python/blob/3f4f84d/cdd/tests/test_docstring/test_parse_docstring_utils.py

---

@fchollet Can I send smaller PRs your way? - Maybe that just fix ambiguities and inconsistencies without introducing the PEP484 [and derived] syntax throughout your codebase?
